### ISSUE TITLE
Virtual Service does not match on other port than `80`

### ISSUE THREAD

**Issue by Kyslik** [istio/istio#38934]
### Bug Description

TLDR: Virtual Service does not match traffic on other port than `80` (and AFAIK it used to work in previous versions, I went through releases and could not find a change that would cause this - from 1.5 -> 1.13, I am unsure when it stopped working 😢 ).

We have following setup:

k8s service:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: a
  namespace: a
spec:
  ports:
  - name: http
    port: 8080
    protocol: TCP
    targetPort: 8080
  - name: http-health
    port: 8079
    protocol: TCP
    targetPort: 8079
  selector:
    app: a
  type: ClusterIP
```

VirtualService:

```yaml
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: a
  namespace: b
spec:
  exportTo:
  - .
  hosts:
  - a.a.svc.cluster.local
  http:
  - match:
    - method:
        exact: GET
      uri:
        exact: /v1/wait
    name: GET-/v1/wait
    retries:
      attempts: 2
      perTryTimeout: 2s
      retryOn: gateway-error,connect-failure,retriable-status-codes
    route:
    - destination:
        host: a.a.svc.cluster.local
        port:
          number: 8080
      headers:
        request:
          add:
            x-envoy-retriable-status-codes: 408,423
    timeout: 6s
  - name: provider-global-default
    retries:
      attempts: 3
      perTryTimeout: 1s
      retryOn: gateway-error,connect-failure,retriable-status-codes
    route:
    - destination:
        host: a.a.svc.cluster.local
        port:
          number: 8080
      headers:
        request:
          add:
            x-envoy-retriable-status-codes: 408,423
    timeout: 10s
```

> Note: the workload `a.a.svc.cluster.local:8080/v1/wait/x` waits for `x` seconds and responds with status code 200 afterwards.

When we curl from namespace `b` to `a.a.svc.cluster.local:8080/v1/wait/30` we get unexpected result, it waits 30 seconds instead of time outing at ~4 second mark.

```sh
time curl -v -X GET "http://a.a.svc.cluster.local:8080/v1/wait/30"
> GET /v1/wait/30 HTTP/1.1
> User-Agent: curl/7.35.0
> Host: a.a.svc.cluster.local:8080
> Accept: */*
>
< HTTP/1.1 200 OK
// SNIP
< server: envoy
<
real	0m 30.04s
user	0m 0.00s
sys	0m 0.00s
```

However when we use either port 80 or don't specify port, everything works as expected:

```sh
time curl -v -X GET "http://a.a.svc.cluster.local:80/v1/wait/30"
> GET /v1/wait/30 HTTP/1.1
> User-Agent: curl/7.35.0
> Host: a.a.svc.cluster.local
> Accept: */*
>
< HTTP/1.1 504 Gateway Timeout
< content-length: 24
< content-type: text/plain
< date: Mon, 16 May 2022 13:00:37 GMT
< server: envoy
<
upstream request timeoutreal	0m 4.13s
user	0m 0.00s
sys	0m 0.00s
```

We also tried adding port to the [HTTPMatchRequest](https://istio.io/latest/docs/reference/config/networking/virtual-service/#HTTPMatchRequest) but no joy.

The proxy-config shows Istio defaults for port `8080`, as for port `80` it seems to apply desired rules from VirtualService shown above. See below:
  
```yaml
- name: "80"
  validateClusters: false
  virtualHosts:
  - domains:
    - a.a.svc.cluster.local
    - a.a.svc.cluster.local:80
    includeRequestAttemptCount: true
    name: a.a.svc.cluster.local:80
    routes:
    - decorator:
        operation: a.a.svc.cluster.local:8080/v1/wait
      match:
        caseSensitive: true
        headers:
        - exactMatch: GET
          name: :method
        path: /v1/wait
      metadata:
        filterMetadata:
          istio:
            config: /apis/networking.istio.io/v1alpha3/namespaces/b/virtual-service/a
      name: GET-/v1/wait
      requestHeadersToAdd:
      - append: true
        header:
          key: x-envoy-retriable-status-codes
          value: 408,423
      route:
        cluster: outbound|8080||a.a.svc.cluster.local
        maxGrpcTimeout: 7s
        retryPolicy:
          hostSelectionRetryMaxAttempts: "5"
          numRetries: 2
          perTryTimeout: 2s
          retryHostPredicate:
          - name: envoy.retry_host_predicates.previous_hosts
          retryOn: gateway-error,connect-failure,retriable-status-codes
        timeout: 6s
    - decorator:
        operation: a.a.svc.cluster.local:8080/*
      match:
        prefix: /
      metadata:
        filterMetadata:
          istio:
            config: /apis/networking.istio.io/v1alpha3/namespaces/b/virtual-service/a
      name: provider-global-default
      requestHeadersToAdd:
      - append: true
        header:
          key: x-envoy-retriable-status-codes
          value: 408,423
      route:
        cluster: outbound|8080||a.a.svc.cluster.local
        maxGrpcTimeout: 10s
        retryPolicy:
          hostSelectionRetryMaxAttempts: "5"
          numRetries: 3
          perTryTimeout: 1s
          retryHostPredicate:
          - name: envoy.retry_host_predicates.previous_hosts
          retryOn: gateway-error,connect-failure,retriable-status-codes
        timeout: 10s
- name: "8080"
  validateClusters: false
  virtualHosts:
  - domains:
    - a.a.svc.cluster.local
    - a.a.svc.cluster.local:8080
    - a.a
    - a.a:8080
    - a.a.svc
    - a.a.svc:8080
    includeRequestAttemptCount: true
    name: a.a.svc.cluster.local:8080
    routes:
    - decorator:
        operation: a.a.svc.cluster.local:8080/*
      match:
        prefix: /
      name: default
      route:
        cluster: outbound|8080||a.a.svc.cluster.local
        maxStreamDuration:
          grpcTimeoutHeaderMax: 0s
          maxStreamDuration: 0s
        retryPolicy:
          hostSelectionRetryMaxAttempts: "5"
          numRetries: 2
          retriableStatusCodes:
          - 503
          retryHostPredicate:
          - name: envoy.retry_host_predicates.previous_hosts
          retryOn: connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes
        timeout: 0s
// SNIP
```

### Version

```prose
$ istioctl version
client version: 1.13.3
control plane version: 1.13.2
data plane version: 1.13.2 (366 proxies)
$ kubectl version --short
Client Version: v1.24.0
Kustomize Version: v4.5.4
Server Version: v1.21.9-eks-0d102a7
```

**Comment by Kyslik** [comment]
I think this is it https://github.com/istio/istio/issues/37691
